### PR TITLE
[TDD] Implementation of DBus Model interface @open sesame 08/03 11:50

### PR DESCRIPTION
**Pull request by again4you**
This patch newly adds the unit test case of the DBus Model interface and its implementation code. DBus Model interface handles the neural network model files by storing the name of a model file and its file path in the internal key-value database. To improve the maintainability, this PR defines the `IMLServiceDB` interface so the backend database could be changed easily.

#### V3
* Use `#define` header-guard instead of `#pragma once` since it is not standard in C++

#### V2
* Move the `client proxy` to SetUp() method.
* Remove unused headers
* Fix the boilerplate code

#### V1
* First draft
* Implementation of DBus Model interface and its test cases

### CODE REVIEW

File: daemon/model-dbus-impl.cc [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+ * @brief DBus implementation for Model Interface
+ * @see	https://github.com/nnstreamer/api
+ * @author Sangjung Woo <sangjung.woo@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+
+#include <glib.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <gst/gst.h>
```
Comment (anyj0527) [inline daemon/model-dbus-impl.cc:16]: Is gst needed?

File: daemon/model-dbus-impl.cc [hunk 2]
```diff
@@ -0,0 +1,269 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/**
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * @file test-dbus-impl.c
```
Comment (gichan-jang) [inline daemon/model-dbus-impl.cc:5]: Please change the file name.
```suggestion
 * @file model-dbus-impl.cc
```

File: tests/daemon/unittest_dbus_model.cc [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    G_BUS_TYPE_SESSION,
+    G_DBUS_PROXY_FLAGS_NONE,
+    DBUS_ML_BUS_NAME,
+    DBUS_MODEL_PATH,
+    NULL, &error);
+  if (error != nullptr) {
+    g_critical ("Error : %s", error->message);
+    g_error_free (error);
+    FAIL();
+  }
```
Comment (gichan-jang) [inline tests/daemon/unittest_dbus_model.cc:92]: These lines are repeated. Is there any reason why they are not included in the test fixture?

File: daemon/model-dbus-impl.cc (hunk 1 above)
Comment (again4you) [reply daemon/model-dbus-impl.cc:16]: You are right! I need to remove unused headers.

File: daemon/model-dbus-impl.cc (hunk 2 above)
Comment (again4you) [reply daemon/model-dbus-impl.cc:5]: Nice catch! 👍

File: tests/daemon/unittest_dbus_model.cc (hunk 3 above)
Comment (again4you) [reply tests/daemon/unittest_dbus_model.cc:92]: My fault. I will move this repeated code to `SetUp()` method.

**Comment by again4you** [comment]
#### V2
* Move the `client proxy` to SetUp() method.
* Remove unused headers
* Fix the boilerplate code

File: daemon/meson.build [hunk 4]
```diff
@@ -13,14 +21,19 @@ if get_option('enable-machine-learning-agent')
               '--generate-c-code', 'pipeline-dbus',
               '--output-directory', meson.current_build_dir(),
               '@INPUT@'])
-
   nns_ml_agent_gen_srcs += gdbus_gen_src
-
-  nns_ml_agent_srcs += join_paths('main.c')
-  nns_ml_agent_srcs += join_paths('modules.c')
-  nns_ml_agent_srcs += join_paths('gdbus-util.c')
   nns_ml_agent_srcs += join_paths('pipeline-module.c')
```
Comment (jaeyun-jung) [inline daemon/meson.build:25]: move this to line 14?

**Review by anyj0527 — APPROVED** [review]
LGTM 👍

File: daemon/meson.build (hunk 4 above)
Comment (again4you) [reply daemon/meson.build]: Applied as below. Thank you for notifying me.
```c++
  nns_ml_agent_srcs += join_paths('gdbus-util.c')
  nns_ml_agent_srcs += join_paths('service-db.cc')
  nns_ml_agent_srcs += join_paths('pipeline-module.c')
  nns_ml_agent_srcs += join_paths('model-dbus-impl.cc')
```

**Review by jaeyun-jung — APPROVED** [review]

File: daemon/includes/service-db.hh [hunk 5]
```diff
@@ -0,0 +1,67 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/**
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * @file service-db.hh
+ * @date 28 Mar 2022
+ * @brief NNStreamer/Service Database Interface
+ * @see	https://github.com/nnstreamer/nnstreamer
```
Comment (wooksong) [inline daemon/includes/service-db.hh:8]:
```suggestion
 * @see	https://github.com/nnstreamer/api
```

File: daemon/service-db.cc [hunk 6]
```diff
@@ -0,0 +1,190 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/**
+ * Copyright (c) 2022 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * @file service-db.cc
+ * @date 29 Jul 2022
+ * @brief Database implementation of NNStreamer/Service C-API
+ * @see https://github.com/nnstreamer/nnstreamer
```
Comment (wooksong) [inline daemon/service-db.cc:8]:
```suggestion
 * @see https://github.com/nnstreamer/api
```

File: daemon/includes/service-db.hh [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ * @author Sangjung Woo <sangjung.woo@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (again4you) [reply daemon/includes/service-db.hh:13]: OK. You are right. I will fix it.

File: daemon/includes/service-db.hh (hunk 5 above)
Comment (again4you) [reply daemon/includes/service-db.hh]: fixed.

File: daemon/service-db.cc (hunk 6 above)
Comment (again4you) [reply daemon/service-db.cc]: fixed.

**Comment by again4you** [comment]
#### V3
* Use `#define` header-guard instead of `#pragma once` since it is not standard in C++

**Review by gichan-jang — APPROVED** [review]
LGTM

**Pull request merged**